### PR TITLE
Drop "disable_max_key_fee"

### PR DESCRIPTION
**Pull request by walidmujahid**
From: https://github.com/lbryio/lbry/issues/929

- [x] Rmeove `disable_max_key_fee` and roll functionality into `max_key_fee`.
- [ ] Test `check_fee_and_convert`

### CODE REVIEW

**Comment by jackrobison** [comment]
Please rebase to master. The docs files are automatically generated from the docstrings - so they shouldn't have any changes to them directly. Otherwise, looks good to me :+1:

**Comment by jackrobison** [comment]
@walidmujahid Upon thinking about it a little more I realized this setting is not redundant. I think it's fine to drop it as a separate setting, but the functionality should be rolled in to the max_key_fee setting (such as disabling by setting it to -1)

**Comment by walidmujahid** [comment]
> @walidmujahid Upon thinking about it a little more I realized this setting is not redundant. I think it's fine to drop it as a separate setting, but the functionality should be rolled in to the max_key_fee setting (such as disabling by setting it to -1)

I see @jackrobison. I can roll the functionality accordingly. Do you want me to do so and then do the rebase to master?

> Thanks for the PR! Question: If a user currently has this setting in their daemon_settings.yml, will it clear it out for them?

I am not sure @tzarebczan 

> Can we send you some LBC in appreciation for your efforts?

I would not say no to that :-) After all issues are fixed and the PR is accepted, do you want me to get with you on this on the discord?

**Comment by jackrobison** [comment]
>I see @jackrobison. I can roll the functionality accordingly. Do you want me to do so and then do the rebase to master?

@walidmujahid :+1:

**Comment by walidmujahid** [comment]
@jackrobison I am not fully sure about the way I went about doing it. Could I get some feedback?

**Comment by walidmujahid** [comment]
> Question: If a user currently has this setting in their daemon_settings.yml, will it clear it out for them?

@tzarebczan I looked into it and it does not seem to clear out automatically when the API server restarts. Rather, the daemon_settings.yml is updated whenever something is changed via `settings_set`.

**Comment by tzarebczan** [comment]
Maybe not automatically clear out all, but for cases like this where we are removing a setting, it may be worth to avoid confusion. I'll let @jackrobison / @eukreign decide how we should handle this/other cases.

File: lbrynet/daemon/Downloader.py [hunk 1]
```diff
@@ -99,12 +97,15 @@ def set_status(self, status, name):
 
     @defer.inlineCallbacks
     def check_fee_and_convert(self, fee):
-        max_key_fee_amount = self.convert_max_fee()
+        currency, amount = self.max_key_fee['currency'], self.max_key_fee['amount']
+        max_key_fee_amount = self.convert_max_fee(currency, amount)
         converted_fee_amount = self.exchange_rate_manager.convert_currency(fee.currency, "LBC",
                                                                            fee.amount)
```
Comment (eukreign) [inline lbrynet/daemon/Downloader.py]: Doesn't the above two lines do the same thing but with different values? Why is `self.convert_max_fee()` called for one but `converted_fee_amount` calculated here? Can we drop `self.convert_max_fee` if it's not needed?

File: lbrynet/daemon/Downloader.py [hunk 2]
```diff
@@ -89,8 +88,7 @@ def check_status(self):
         else:
             log.debug("Waiting for stream descriptor (%i seconds)", self.timeout_counter)
 
-    def convert_max_fee(self):
-        currency, amount = self.max_key_fee['currency'], self.max_key_fee['amount']
+    def convert_max_fee(self, currency, amount):
         return self.exchange_rate_manager.convert_currency(currency, "LBC", amount)
```
Comment (eukreign) [inline lbrynet/daemon/Downloader.py]: Why these changes?

File: lbrynet/daemon/Downloader.py [hunk 3]
```diff
@@ -99,12 +97,15 @@ def set_status(self, status, name):
...
+        currency, amount = self.max_key_fee['currency'], self.max_key_fee['amount']
+        max_key_fee_amount = self.convert_max_fee(currency, amount)
         converted_fee_amount = self.exchange_rate_manager.convert_currency(fee.currency, "LBC",
                                                                            fee.amount)
+        is_disabled_max_key_fee = True if amount < 0 else False
+
         if converted_fee_amount > (yield self.wallet.default_account.get_balance()):
             raise InsufficientFundsError('Unable to pay the key fee of %s' % converted_fee_amount)
-        if converted_fee_amount > max_key_fee_amount and not self.disable_max_key_fee:
+        if converted_fee_amount > max_key_fee_amount and not is_disabled_max_key_fee:
```
Comment (eukreign) [inline lbrynet/daemon/Downloader.py]: This is really verbose and redundant:
```
is_disabled_max_key_fee = True if amount < 0 else False
```
above can be represented much more succinctly as:
```
is_disabled_max_key_fee = amount < 0
```
but since the `is_disabled_max_key_fee` isn't used anywhere else and is making a double negative `(not amount < 0)` you could make this much cleaner and easier to read by removing the `not` and flipping the less than to greater than:
```
 if converted_fee_amount > max_key_fee_amount and amount > 0:
```
unless I'm missing something in the logic...

Comment (walidmujahid) [reply lbrynet/daemon/Downloader.py]: No, you are not missing anything. It makes complete sense.

File: lbrynet/daemon/Downloader.py (hunk 1 above)
Comment (walidmujahid) [reply lbrynet/daemon/Downloader.py]: >Doesn't the above two lines do the same thing but with different values?

By "above two lines", are you referring to the newly inserted lines, or are you referrring to the use of `self.convert_max_fee` and what `converted_fee_amount` is doing?

>Why is self.convert_max_fee() called for one but converted_fee_amount calculated here? 

Are you asking why `self.convert_max_fee` is not being used for `converted_fee_amount`? If so, I agree. I originally was not paying attention to that in order to keep as close to the way it was just with my changes included. 

That probably was not the best way to go about it. I can change it accordingly. 

>Can we drop self.convert_max_fee if it's not needed?

I would not say it is not needed. It can be used for both `max_key_fee_amount`  and `converted_fee_amount`.

File: lbrynet/daemon/Downloader.py (hunk 2 above)
Comment (walidmujahid) [reply lbrynet/daemon/Downloader.py]: I had something else in mind at the time but based on an earlier comment of yours, I would say that this change would allow for the method's use for `converted_fee_amount`.

File: lbrynet/daemon/Downloader.py [hunk 4]
```diff
@@ -89,22 +88,21 @@ def check_status(self):
...
         log.info("Download lbry://%s status changed to %s" % (name, status))
         self.code = next(s for s in STREAM_STAGES if s[0] == status)
 
     @defer.inlineCallbacks
     def check_fee_and_convert(self, fee):
-        max_key_fee_amount = self.convert_max_fee()
-        converted_fee_amount = self.exchange_rate_manager.convert_currency(fee.currency, "LBC",
-                                                                           fee.amount)
+        max_key_fee_amount = self.convert_max_fee(self.max_key_fee)
+        converted_fee_amount = self.convert_max_fee(fee.currency, fee.amount)
```
Comment (eukreign) [inline lbrynet/daemon/Downloader.py]: Can't see how this could work. Did you run this code?

Comment (walidmujahid) [reply lbrynet/daemon/Downloader.py]: Fixed it. Wrong version.

File: tests/unit/lbrynet_daemon/test_Downloader.py [hunk 5]
```diff
@@ -103,6 +106,17 @@ def test_init_exception(self):
         with self.assertRaises(AttributeError):
             yield getstream.start(stream_info, name, "deadbeef" * 12, 0)
 
+    @defer.inlineCallbacks
+    def test_getstream_convert_fee_method(self):
+        getstream = self.init_getstream_with_mocs()
+        name = 'test'
+        stream_info = None
+
+        yield self.assertTrue(moc_convert_fee("USD", 60))
```
Comment (eukreign) [inline tests/unit/lbrynet_daemon/test_Downloader.py]: What is this line testing?

File: tests/unit/lbrynet_daemon/test_Downloader.py [hunk 6]
```diff
@@ -103,6 +106,17 @@ def test_init_exception(self):
...
+        yield self.assertRaises(TypeError, getstream.convert_fee, "USD:60")
+        yield self.assertRaises(TypeError, getstream.convert_fee,
+                                "USD:60", "USD", 60)
```
Comment (eukreign) [inline tests/unit/lbrynet_daemon/test_Downloader.py]: `assertTrue` and `assertRaises` do not return `Deferreds` so I'm not sure if these test even runs, did you try running the test? Put an `self.assertTrue(False)` at the end of the test, if the test doesn't fail then it's not getting run :-p

Comment (walidmujahid) [reply tests/unit/lbrynet_daemon/test_Downloader.py]: Yep, I ran the test. It did indeed run locally for me and it seems Travis ran it too. 

Maybe I am not fully understanding what you are saying.

**Comment by tzarebczan** [comment]
@walidmujahid is this ready to be reviewed again? I re-assigned @eukreign to take a peek.

**Comment by walidmujahid** [comment]
> @walidmujahid is this ready to be reviewed again? I re-assigned @eukreign to take a peek.

Not yet @tzarebczan. I am still struggling a bit with the tests, but I hope to have done as soon as possible.

**Comment by alyssaoc** [comment]
try to re-establish contact, if not close

**Comment by walidmujahid** [comment]
Still working on tests @alyssaoc.

**Comment by eukreign** [comment]
The code touched by this PR has changed significantly.

**Comment by eukreign** [comment]
Dropped `--disable-max-key-fee` as a separately stored boolean and instead there is just `max_key_fee: null` in the daemon_settings.yml, since setting things null via command line is difficult there is new `--no-max-key-fee` which sets `max_key_fee: null`. Also, passing a new max key fee via cli is now using this format: `--max-key-fee 3.0 USD` (in daemon_settings.yml it remains unchanged for backwards compatibility)

https://github.com/lbryio/lbry/pull/1815
